### ISSUE TITLE
`builtins.filter` support for PEP 647 `TypeGuard`

### ISSUE THREAD

**Issue by Jasha10** [microsoft/pyright#2768]
**Is your feature request related to a problem? Please describe.**
Currently it seems that pyright does not support passing a type guard function as the first argument to the builtin `filter` function.
Here is an example:
```python
# tmp.py
from typing import Iterable, Optional, TypeGuard


def is_not_none(x: Optional[int]) -> TypeGuard[int]:
    return x is not None

generate_optional: Iterable[Optional[int]] = (x if x % 2 == 0 else None for x in range(100))
list_optional: list[Optional[int]] = [0, None, 1, None, 2]
generate_ints: Iterable[int] = filter(is_not_none, list_optional)
```
```text
$ pyright tmp.py
...
  /home/jasha10/tmp.py:9:32 - error: Expression of type "filter[int | None]" cannot be assigned to declared type "Iterable[int]"
    TypeVar "_T_co@Iterable" is covariant
      Type "int | None" cannot be assigned to type "int"
        Type "None" cannot be assigned to type "int" (reportGeneralTypeIssues)
1 error, 0 warnings, 0 infos
```

**Describe the solution you'd like**
It would be nice if type guard functions could be used with functional programming primitives such as `filter`.

**Comment by erictraut** [comment]
This isn't a missing feature in pyright. It's a missing overload in the `builtins.pyi` type stub for the `filter` class. 

If you add the following overload signature to the `filter` class, your example will type check fine.

```python
    @overload
    def __init__(
        self, __function: Callable[[_S], TypeGuard[_T]], __iterable: Iterable[_S]
    ) -> None: ...
```

You can submit a feature request — or better yet, a PR — in the typeshed github repo. Pyright picks up new versions of typeshed stubs frequently.

**Comment by Jasha10** [comment]
Excellent, thanks.

**Comment by layday** [comment]
Another PR to add this to the typeshed was closed last week because it confused mypy, see https://github.com/python/typeshed/pull/6140.
